### ISSUE TITLE
Problem in building serialbox2 with gridtools unittests

### ISSUE THREAD

**Issue by adamryczkowski** [eth-cscs/serialbox2#11]
```
[  8%] Built target gtest
[ 30%] Built target SerialboxCoreLibraryObjects
[ 31%] Built target SerialboxCoreLibraryShared
[ 32%] Built target SerialboxCoreLibraryStatic
[ 41%] Built target SerialboxCLibraryObjects
[ 42%] Built target SerialboxCLibraryStatic
[ 42%] Built target sdbPython
[ 42%] Built target SerialboxPython
[ 43%] Built target SerialboxCLibraryShared
[ 47%] Built target SerialboxUnittestUtility
[ 48%] Building CXX object test/serialbox/core/CMakeFiles/SerialboxCppUnittest.dir/frontend/gridtools/UnittestStorageView.cpp.o
In file included from /home/adam/tmp/gridtools/include/storage/meta_storage.hpp:39:0,
                 from /home/adam/tmp/gridtools/include/storage/meta_storage_extender.hpp:37,
                 from /home/adam/tmp/gridtools/include/storage/storage.hpp:51,
                 from /home/adam/tmp/gridtools/include/storage/storage_metafunctions.hpp:44,
                 from /home/adam/tmp/gridtools/include/storage/storage-facility.hpp:38,
                 from /home/adam/tmp/serialbox2/test/utility/GridTools.h:27,
                 from /home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:15:
/home/adam/tmp/serialbox2/src/serialbox/core/frontend/gridtools/StorageViewHelper.h: In instantiation of ‘std::vector<int, std::allocator<int> > serialbox::gridtools::internal::get_strides(const StorageType&, const MetaDataType&) [with MetaDataType = gridtools::meta_storage<gridtools::meta_storage_aligned<gridtools::meta_storage_base<mpl_::integral_c<unsigned int, 1u>, gridtools::layout_map<0, 1>, false>, gridtools::aligned<0u>, gridtools::halo<1u, 2u> > >; StorageType = gridtools::storage<gridtools::base_storage<gridtools::wrap_pointer<int, true>, gridtools::meta_storage<gridtools::meta_storage_aligned<gridtools::meta_storage_base<mpl_::integral_c<unsigned int, 1u>, gridtools::layout_map<0, 1>, false>, gridtools::aligned<0u>, gridtools::halo<1u, 2u> > >, 1u> >]’:
/home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:181:3:   required from ‘void GridToolsStorageViewTest_Construction_Test<gtest_TypeParam_>::TestBody() [with gtest_TypeParam_ = int]’
/home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:485:1:   required from here
/home/adam/tmp/gridtools/include/storage/meta_storage_base.hpp:134:42: error: ‘gridtools::array<int, 2ul> gridtools::meta_storage_base<mpl_::integral_c<unsigned int, 1u>, gridtools::layout_map<0, 1>, false>::m_strides’ is protected
         array< int_t, space_dimensions > m_strides;
                                          ^
In file included from /home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:23:0:
/home/adam/tmp/serialbox2/src/serialbox/core/frontend/gridtools/StorageViewHelper.h:43:41: error: within this context
   const auto& strides_array = meta_data.m_strides;
                                         ^
In file included from /home/adam/tmp/gridtools/include/storage/meta_storage.hpp:39:0,
                 from /home/adam/tmp/gridtools/include/storage/meta_storage_extender.hpp:37,
                 from /home/adam/tmp/gridtools/include/storage/storage.hpp:51,
                 from /home/adam/tmp/gridtools/include/storage/storage_metafunctions.hpp:44,
                 from /home/adam/tmp/gridtools/include/storage/storage-facility.hpp:38,
                 from /home/adam/tmp/serialbox2/test/utility/GridTools.h:27,
                 from /home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:15:
/home/adam/tmp/serialbox2/src/serialbox/core/frontend/gridtools/StorageViewHelper.h: In instantiation of ‘void* serialbox::gridtools::internal::get_origin_ptr(const StorageType&, const MetaDataType&, unsigned int) [with StorageType = gridtools::storage<gridtools::base_storage<gridtools::wrap_pointer<int, true>, gridtools::meta_storage<gridtools::meta_storage_aligned<gridtools::meta_storage_base<mpl_::integral_c<unsigned int, 1u>, gridtools::layout_map<0, 1>, false>, gridtools::aligned<0u>, gridtools::halo<1u, 2u> > >, 1u> >; MetaDataType = gridtools::meta_storage<gridtools::meta_storage_aligned<gridtools::meta_storage_base<mpl_::integral_c<unsigned int, 1u>, gridtools::layout_map<0, 1>, false>, gridtools::aligned<0u>, gridtools::halo<1u, 2u> > >]’:
/home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:181:3:   required from ‘void GridToolsStorageViewTest_Construction_Test<gtest_TypeParam_>::TestBody() [with gtest_TypeParam_ = int]’
/home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:485:1:   required from here
/home/adam/tmp/gridtools/include/storage/meta_storage_base.hpp:134:42: error: ‘gridtools::array<int, 2ul> gridtools::meta_storage_base<mpl_::integral_c<unsigned int, 1u>, gridtools::layout_map<0, 1>, false>::m_strides’ is protected
         array< int_t, space_dimensions > m_strides;
                                          ^
In file included from /home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:23:0:
/home/adam/tmp/serialbox2/src/serialbox/core/frontend/gridtools/StorageViewHelper.h:91:41: error: within this context
   const auto& strides_array = meta_data.m_strides;
                                         ^
In file included from /home/adam/tmp/gridtools/include/storage/meta_storage.hpp:39:0,
                 from /home/adam/tmp/gridtools/include/storage/meta_storage_extender.hpp:37,
                 from /home/adam/tmp/gridtools/include/storage/storage.hpp:51,
                 from /home/adam/tmp/gridtools/include/storage/storage_metafunctions.hpp:44,
                 from /home/adam/tmp/gridtools/include/storage/storage-facility.hpp:38,
                 from /home/adam/tmp/serialbox2/test/utility/GridTools.h:27,
                 from /home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:15:
/home/adam/tmp/serialbox2/src/serialbox/core/frontend/gridtools/StorageViewHelper.h: In instantiation of ‘std::vector<int, std::allocator<int> > serialbox::gridtools::internal::get_strides(const StorageType&, const MetaDataType&) [with MetaDataType = gridtools::meta_storage<gridtools::meta_storage_aligned<gridtools::meta_storage_base<mpl_::integral_c<unsigned int, 3u>, gridtools::layout_map<0, 1, -1>, false>, gridtools::aligned<0u>, gridtools::halo<1u, 2u, 3u> > >; StorageType = gridtools::storage<gridtools::base_storage<gridtools::wrap_pointer<int, true>, gridtools::meta_storage<gridtools::meta_storage_aligned<gridtools::meta_storage_base<mpl_::integral_c<unsigned int, 3u>, gridtools::layout_map<0, 1, -1>, false>, gridtools::aligned<0u>, gridtools::halo<1u, 2u, 3u> > >, 1u> >]’:
/home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:201:3:   required from ‘void GridToolsStorageViewTest_Construction_Test<gtest_TypeParam_>::TestBody() [with gtest_TypeParam_ = int]’
/home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:485:1:   required from here
/home/adam/tmp/gridtools/include/storage/meta_storage_base.hpp:134:42: error: ‘gridtools::array<int, 3ul> gridtools::meta_storage_base<mpl_::integral_c<unsigned int, 3u>, gridtools::layout_map<0, 1, -1>, false>::m_strides’ is protected
         array< int_t, space_dimensions > m_strides;
                                          ^
In file included from /home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:23:0:
/home/adam/tmp/serialbox2/src/serialbox/core/frontend/gridtools/StorageViewHelper.h:43:41: error: within this context
   const auto& strides_array = meta_data.m_strides;
                                         ^
In file included from /home/adam/tmp/gridtools/include/storage/meta_storage.hpp:39:0,
                 from /home/adam/tmp/gridtools/include/storage/meta_storage_extender.hpp:37,
                 from /home/adam/tmp/gridtools/include/storage/storage.hpp:51,
                 from /home/adam/tmp/gridtools/include/storage/storage_metafunctions.hpp:44,
                 from /home/adam/tmp/gridtools/include/storage/storage-facility.hpp:38,
                 from /home/adam/tmp/serialbox2/test/utility/GridTools.h:27,
                 from /home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:15:
/home/adam/tmp/serialbox2/src/serialbox/core/frontend/gridtools/StorageViewHelper.h: In instantiation of ‘void* serialbox::gridtools::internal::get_origin_ptr(const StorageType&, const MetaDataType&, unsigned int) [with StorageType = gridtools::storage<gridtools::base_storage<gridtools::wrap_pointer<int, true>, gridtools::meta_storage<gridtools::meta_storage_aligned<gridtools::meta_storage_base<mpl_::integral_c<unsigned int, 3u>, gridtools::layout_map<0, 1, -1>, false>, gridtools::aligned<0u>, gridtools::halo<1u, 2u, 3u> > >, 1u> >; MetaDataType = gridtools::meta_storage<gridtools::meta_storage_aligned<gridtools::meta_storage_base<mpl_::integral_c<unsigned int, 3u>, gridtools::layout_map<0, 1, -1>, false>, gridtools::aligned<0u>, gridtools::halo<1u, 2u, 3u> > >]’:
/home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:201:3:   required from ‘void GridToolsStorageViewTest_Construction_Test<gtest_TypeParam_>::TestBody() [with gtest_TypeParam_ = int]’
/home/adam/tmp/serialbox2/test/serialbox/core/frontend/gridtools/UnittestStorageView.cpp:485:1:   required from here
/home/adam/tmp/gridtools/include/storage/meta_storage_base.hpp:134:42: error: ‘gridtools::array<int, 3ul> gridtools::meta_storage_base<mpl_::integral_c<unsigned int, 3u>, gridtools::layout_map<0, 1, -1>, false>::m_strides’ is protected
         array< int_t, space_dimensions > m_strides;
                                          ^
```

Steps to reproduce:

Install vanilla Ubuntu 16.04 server and then
```
sudo apt install cmake build-essential gfortran git libboost-filesystem-dev libboost-system-dev libboost-chrono-dev libboost-log-dev
git clone --depth 1 https://github.com/eth-cscs/serialbox2.git
git clone --depth 1 git@github.com:GridTools/gridtools.git
mkdir serialbox2/build
cd serialbox2/build
cmake -DSERIALBOX_TESTING=ON -DSERIALBOX_TESTING_GRIDTOOLS=ON -DSERIALBOX_GRIDTOOLS_ROOT=../../gridtools -DGRIDTOOLS_ROOT=../../gridtools ..
make
```

**Comment by adamryczkowski** [comment]
I use up-to-date gridtools with last commit
```
commit 87d9ca198cd015b4c3f61d2c000c52fe67e8ff2e
Author: Hannes Vogt <hannes@havogt.de>
Date:   Fri Mar 10 11:19:01 2017 +0100

    Fix daint environment (#594)
    
    Bug description: The slurm partition was not set correctly in the slurm script and the timelimit was to strict.
    
    Changes: Increase timelimit to 15 min (instead of 10 min).
```

**Comment by thfabian** [comment]
Hey Adam,
It is unlikely that the latest gridtools commit works as I activly developed the library in late 2016 and gridtools is a rapidly progressing library. I'll fix it as soon as I can get my hands on a computer. You will have it by noon.

**Comment by adamryczkowski** [comment]
Thanks! It is not that urgent. For me it can wait at least 2 work days. I started work on incorporation the Fortran part to our project.

**Comment by thfabian** [comment]
The problem was that my original PR to allow serialbox2 to compile was never merged upstream. I created a new PR (https://github.com/eth-cscs/gridtools/pull/629). In the mean time you can pull from my fork (https://github.com/thfabian/gridtools/tree/serialbox2_fix) branch `serialbox2_fix`.

Let me know if there are any further problems.

**Comment by adamryczkowski** [comment]
@thfabian Do you know why your patch has not been merged into gridtools? Is it safe to use your fork of gridtools?

**Comment by adamryczkowski** [comment]
What happened to the https://github.com/eth-cscs/gridtools anyway? It doesn't seem to be on Github anymore...

**Comment by thfabian** [comment]
gridtools underwent a major update of the storage infrastructure and thus the gridtools frontend of serialbox needs to some adjustments to function properly again. @havogt will take over development of serialbox but in the mean time you can use my fork (the changes are very minimal anyway).

**Comment by adamryczkowski** [comment]
The fork you mentioned still doesn't exist, (or maybe it is not available to me).

**Comment by havogt** [comment]
Yes, eth-cscs/gridtools is the internal repository. We will update GridTools/gridtools and make the fix available to you.

**Comment by havogt** [comment]
This issue is resolved with gridtools 1.05.01.
